Building test suite fails in a different way on Windows 10 (zlib build)

This _appears_ to be different from issue #13 of a similar name in that this build err on Win 10 occurs during the zlib build.  Teh google suggests this is a result of a dependency on the MSVC runtime library on a Cygwin-equipped box.

Should I simply install an MSVC runtime env and be done with it? I'm not sure which version.  TIA Howard

[edited to show errs exposed built within a cygwin env, orig post exhibited same error regarding missing _wopen but without benefit (?) of the cygwin environment]

```
[ 55%] Building C object CMakeFiles/zlib.dir/inftrees.o
[ 60%] Building C object CMakeFiles/zlib.dir/inffast.o
[ 65%] Building C object CMakeFiles/zlib.dir/trees.o
/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib/gzlib.c: In function ‘gz_open’:
/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib/gzlib.c:243:20: warning: implicit declaration of function ‘_wopen’; did you mean ‘fwopen’? [-Wimplicit-function-declaration]
         fd == -2 ? _wopen(path, oflag, 0666) :
                    ^~~~~~
                    fwopen
[ 70%] Building C object CMakeFiles/zlib.dir/uncompr.o
[ 75%] Building C object CMakeFiles/zlib.dir/zutil.o
[ 80%] Linking C static library libz.a
make[8]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
[ 80%] Built target zlib
make[8]: Entering directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[8]: Entering directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
Scanning dependencies of target example
make[8]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
Scanning dependencies of target minigzip
make[8]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[8]: Entering directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[8]: Entering directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
[ 85%] Building C object CMakeFiles/example.dir/test/example.o
[ 90%] Building C object CMakeFiles/minigzip.dir/test/minigzip.o
[ 95%] Linking C executable example.exe
[100%] Linking C executable minigzip.exe
libz.a(gzlib.o):gzlib.c:(.text+0x3e0): undefined reference to `_wopen'
libz.a(gzlib.o):gzlib.c:(.text+0x3e0): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `_wopen'
libz.a(gzlib.o):gzlib.c:(.text+0x3e0): undefined reference to `_wopen'
libz.a(gzlib.o):gzlib.c:(.text+0x3e0): relocation truncated to fit: R_X86_64_PC32 against undefined symbol `_wopen'
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[8]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[8]: *** [CMakeFiles/example.dir/build.make:96: example.exe] Error 1
make[8]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[8]: *** [CMakeFiles/minigzip.dir/build.make:96: minigzip.exe] Error 1
make[7]: *** [CMakeFiles/Makefile2:68: CMakeFiles/example.dir/all] Error 2
make[7]: *** Waiting for unfinished jobs....
make[7]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[7]: *** [CMakeFiles/Makefile2:142: CMakeFiles/minigzip.dir/all] Error 2
make[6]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib/zlib/src/zlib-build'
make[6]: *** [Makefile:139: all] Error 2
make[5]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib'
make[5]: *** [CMakeFiles/zlib.dir/build.make:111: zlib/src/zlib-stamp/zlib-build] Error 2
make[4]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib'
make[4]: *** [CMakeFiles/Makefile2:68: CMakeFiles/zlib.dir/all] Error 2
make[3]: Leaving directory '/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib'
make[3]: *** [Makefile:84: all] Error 2
CMake Error at third_party/cmake/build_target.cmake:54 (message):

  /cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/third_party/zlib
  failed to build!
Call Stack (most recent call first):
  third_party/CMakeLists.txt:14 (build_target)


-- Configuring incomplete, errors occurred!
See also "/cygdrive/c/Users/HowardGoldstein/Documents/mavlink-testing-suite-master/build/build_mavsdk/CMakeFiles/CMakeOutput.log".
make[2]: *** [CMakeFiles/third_party_mavsdk.dir/build.make:106: third_party_mavsdk-prefix/src/third_party_mavsdk-stamp/third_party_mavsdk-configure] Error 1
make[1]: *** [CMakeFiles/Makefile2:106: CMakeFiles/third_party_mavsdk.dir/all] Error 2
make: *** [Makefile:84: all] Error 2


```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -15,8 +15,10 @@ jobs:
     steps:
     - uses: actions/checkout@v2
     - name: Install dependencies
-      run: sudo apt-get update && sudo apt-get install -y build-essential cmake
+      run: sudo apt-get update && sudo apt-get install -y build-essential cmake clang-format
     - name: Configure
       run: cmake -j 2 -Bbuild -H.
     - name: Build
       run: cmake --build build -j 2
+    - name: Run clang-format
+      run: cmake --build build --target format

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -1,0 +1,22 @@
+name: Build
+
+on:
+  push:
+    branches:
+    - 'master'
+  pull_request:
+    branches:
+    - '*'
+
+jobs:
+  ubuntu-1804:
+    name: Ubuntu-18.04
+    runs-on: ubuntu-18.04
+    steps:
+    - uses: actions/checkout@v2
+    - name: Install dependencies
+      run: sudo apt-get update && sudo apt-get install -y build-essential cmake
+    - name: Configure
+      run: cmake -j 2 -Bbuild -H.
+    - name: Build
+      run: cmake --build build -j 2

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -22,3 +22,13 @@ jobs:
       run: cmake --build build -j 2
     - name: Run clang-format
       run: cmake --build build --target format
+
+  windows:
+    name: Windows
+    runs-on: windows-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Configure
+        run: cmake -G "Visual Studio 16 2019" -j 2 -Bbuild -S.
+      - name: Build
+        run: cmake --build build -j 2 --config Debug

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,5 +1,0 @@
-language: cpp
-before_install:
-install:
-script: ./run_ci.sh
-

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -49,9 +49,9 @@ add_dependencies(mavlink_testing_suite
 
 if(MSVC)
     if (CMAKE_BUILD_TYPE STREQUAL "Release")
-        set (YAML_LIB yaml-cppmd)
+        set (YAML_LIB libyaml-cppmd.lib)
     else()
-        set (YAML_LIB yaml-cppmdd)
+        set (YAML_LIB libyaml-cppmdd.lib)
     endif()
 else()
     set (YAML_LIB yaml-cpp)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -22,8 +22,8 @@ else()
 endif()
 string(TIMESTAMP SUITE_VERSION_BUILD_TIMESTAMP "%Y-%m-%dT%H:%M:%S.000000Z" UTC)
 
-# Specify at least C++11
-set(CMAKE_CXX_STANDARD 11)
+# Specify C++17
+set(CMAKE_CXX_STANDARD 17)
 
 # Enable strict handling of warnings
 if(MSVC)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,17 +8,17 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 find_package(Git)
 
 if (GIT_FOUND)
-	execute_process(
-		COMMAND ${GIT_EXECUTABLE} rev-parse HEAD
-		WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-		OUTPUT_VARIABLE SUITE_VERSION_GIT_REVISION
-		OUTPUT_STRIP_TRAILING_WHITESPACE)
+    execute_process(
+        COMMAND ${GIT_EXECUTABLE} rev-parse HEAD
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE SUITE_VERSION_GIT_REVISION
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
 
-	message(STATUS "Building in git revision ${SUITE_VERSION_GIT_REVISION}")
-	# TODO: build into binary
+    message(STATUS "Building in git revision ${SUITE_VERSION_GIT_REVISION}")
+    # TODO: build into binary
 else()
-	set(SUITE_VERSION_GIT_REVISION "n/a")
-	message(STATUS "Executable git is missing, choosing default value for git revision")
+    set(SUITE_VERSION_GIT_REVISION "n/a")
+    message(STATUS "Executable git is missing, choosing default value for git revision")
 endif()
 string(TIMESTAMP SUITE_VERSION_BUILD_TIMESTAMP "%Y-%m-%dT%H:%M:%S.000000Z" UTC)
 
@@ -47,47 +47,57 @@ add_dependencies(mavlink_testing_suite
     ${dependencies}
     )
 
+if(MSVC)
+    if (CMAKE_BUILD_TYPE STREQUAL "Release")
+        set (YAML_LIB yaml-cppmd)
+    else()
+        set (YAML_LIB yaml-cppmdd)
+    endif()
+else()
+    set (YAML_LIB yaml-cpp)
+endif()
+
 target_link_libraries(mavlink_testing_suite
     mavsdk
     mavsdk_action
-	mavsdk_mission
-	mavsdk_param
-	mavsdk_camera
-	mavsdk_mavlink_passthrough
-	${CMAKE_CURRENT_BINARY_DIR}/build_yaml/libyaml-cpp.a
+    mavsdk_mission
+    mavsdk_param
+    mavsdk_camera
+    mavsdk_mavlink_passthrough
+    ${YAML_LIB}
 )
 
 find_program(CLANG_FORMAT_EXECUTABLE
-	NAMES clang-format clang-format-5.0
-		clang-format-4.0 clang-format-3.9
-		clang-format-3.8 clang-format-3.7
-		clang-format-3.6 clang-format-3.5
-		clang-format-3.4 clang-format-3.3
-	DOC "clang-format executable"
+    NAMES clang-format clang-format-5.0
+        clang-format-4.0 clang-format-3.9
+        clang-format-3.8 clang-format-3.7
+        clang-format-3.6 clang-format-3.5
+        clang-format-3.4 clang-format-3.3
+    DOC "clang-format executable"
 )
 
 file(GLOB_RECURSE MAVLINK_SUITE_FORMAT_SRCS
-	src/*.h src/*.cpp)
+    src/*.h src/*.cpp)
 if (CLANG_FORMAT_EXECUTABLE)
-	message(STATUS "Enabling format target")
-	add_custom_target(
+    message(STATUS "Enabling format target")
+    add_custom_target(
         format ${CMAKE_CURRENT_SOURCE_DIR}/tools/run_clang_format.sh ${CLANG_FORMAT_EXECUTABLE} ${MAVLINK_SUITE_FORMAT_SRCS})
 else()
     message(STATUS "Not enabling format target (clang-format not found)")
 endif()
 
 find_program(CLANG_TIDY_EXECUTABLE
-	NAMES clang-tidy clang-tidy-5.0
-		clang-tidy-4.0 clang-tidy-3.9
-		clang-tidy-3.8 clang-tidy-3.7
-		clang-tidy-3.6 clang-tidy-3.5
-		clang-tidy-3.4 clang-tidy-3.3
-	DOC "clang-tidy executable"
+    NAMES clang-tidy clang-tidy-5.0
+        clang-tidy-4.0 clang-tidy-3.9
+        clang-tidy-3.8 clang-tidy-3.7
+        clang-tidy-3.6 clang-tidy-3.5
+        clang-tidy-3.4 clang-tidy-3.3
+    DOC "clang-tidy executable"
 )
 if (CLANG_TIDY_EXECUTABLE)
-	message(STATUS "Enabling clang-tidy target")
-	add_custom_target(
-		clang_tidy ${CLANG_TIDY_EXECUTABLE} -fix -header-filter="${CMAKE_CURRENT_LIST_DIR}/src/.*\\.[hpp|h]" -p ${CMAKE_CURRENT_BINARY_DIR} ${MAVLINK_SUITE_FORMAT_SRCS})
+    message(STATUS "Enabling clang-tidy target")
+    add_custom_target(
+        clang_tidy ${CLANG_TIDY_EXECUTABLE} -fix -header-filter="${CMAKE_CURRENT_LIST_DIR}/src/.*\\.[hpp|h]" -p ${CMAKE_CURRENT_BINARY_DIR} ${MAVLINK_SUITE_FORMAT_SRCS})
 else()
     message(STATUS "Not enabling clang-tidy target (clang-tidy not found)")
 endif()

--- a/README.md
+++ b/README.md
@@ -4,10 +4,17 @@
 
 This project aims to provide a testing suite to test standard compliance of MAVLink-enabled components/systems. The definition of MAVLink as standard consists of all the messages and only the messages of this message set: https://mavlink.io/en/messages/common.html
 
-- Build the testing suite:
-  ```
-  (mkdir -p build && cd build && cmake .. && make)
-  ```
+Build the testing suite, on Linux/macOS:
+```
+  cmake -Bbuild -S.
+  cmake --build build -j8.
+```
+Or on Windows:
+```
+cmake -G "Visual Studio 16 2019" -Bbuild -S.
+cmake --build build -j8
+```
+
 ### Running the Autopilot Tests
 
 1. Start the SITL simulation (this is specific to the system-under-test, PX4 SITL is simply provided as an example):
@@ -17,7 +24,7 @@ This project aims to provide a testing suite to test standard compliance of MAVL
 
 2. Then run the tests in the `build` directory (this is the same for any tested system):
     ```
-    (cd build && ./mavlink_testing_suite ../config/autopilot.yaml udp://)
+    build/mavlink_testing_suite config/autopilot.yaml udp://
     ```
 
 ### Running the Camera-Manager Tests
@@ -33,7 +40,7 @@ This project aims to provide a testing suite to test standard compliance of MAVL
 
 3. Then run the tests in the `build` directory (this is the same for any tested system):
     ```
-    (cd build && ./mavlink_testing_suite ../config/camera-manager.yaml udp://:14550)
+    build/mavlink_testing_suite ../config/camera-manager.yaml udp://:14550
     ```
 
 **Note:**
@@ -43,7 +50,8 @@ The `camera_id` in [camera-manager.yaml](config/camera-manager.yaml) might need 
 
 Before committing, make sure to run the code formatting and tidying using clang.
 ```
-(cd build && make format; make clang_tidy)
+cmake --build build --target clang_tidy
+cmake --build build --target clang_format
 ```
 
 ## Use custom MAVSDK build
@@ -51,12 +59,7 @@ Before committing, make sure to run the code formatting and tidying using clang.
 For debugging purposes it can be handy to use a custom version of MAVSDK:
 
 ```
-cd <wherever>/MAVSDK
-make clean
-export DRONECODE_SDK_INSTALL_DIR=$(pwd)/install
-make ENABLE_MAVLINK_PASSTHROUGH=1 INSTALL_PREFIX=$MAVSDK_INSTALL_DIR default install
-cd <wherever>/mavlink-testing-suite
-(mkdir -p build && cd build && cmake -DDRONECODE_SDK_INSTALL_DIR=$MAVSDK_INSTALL_DIR .. && make)
+cmake -Bbuild -DMAVSDK_INSTALL_DIR=path/to/MAVSDK/install -S. && cmake --build build -j8)
 ```
 
 ### Sample Output of the Autopilot Tests
@@ -340,4 +343,3 @@ In the test above some errors are logged but can be ignored because there is no 
 [09:41:36|Warn ] sending again, retries to do: 1  (528). (mavlink_commands.cpp:222)
 [09:41:37|Error] Retrying failed (528) (mavlink_commands.cpp:239)
 ```
-

--- a/README.md
+++ b/README.md
@@ -1,6 +1,6 @@
 ## MAVLink Standard Testing Suite
 
-[![Build Status](https://travis-ci.org/Auterion/mavlink-testing-suite.svg?branch=master)](https://travis-ci.org/Auterion/mavlink-testing-suite)
+[![GitHub Actions Status](https://github.com/mavlink/MAVSDK/workflows/Build/badge.svg?branch=master)](https://github.com/mavlink/MAVSDK/actions?query=branch%3Amaster)
 
 This project aims to provide a testing suite to test standard compliance of MAVLink-enabled components/systems. The definition of MAVLink as standard consists of all the messages and only the messages of this message set: https://mavlink.io/en/messages/common.html
 

--- a/cmake/dependencies.cmake
+++ b/cmake/dependencies.cmake
@@ -32,8 +32,9 @@ ExternalProject_Add(third_party_yaml
     SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/third_party/yaml-cpp
     CMAKE_ARGS
         -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}
+       -DYAML_CPP_BUILD_TESTS=OFF
     GIT_REPOSITORY https://github.com/jbeder/yaml-cpp.git
-    GIT_TAG yaml-cpp-0.6.2
+    GIT_TAG yaml-cpp-0.6.3
     BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/build_yaml
     INSTALL_COMMAND ""
     )

--- a/cmake/dependencies.cmake
+++ b/cmake/dependencies.cmake
@@ -8,12 +8,12 @@ if(MAVSDK_INSTALL_DIR)
 else()
     # clone and build MAVSDK via ExternalProject
     ExternalProject_Add(third_party_mavsdk
-        SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/third_party/MAVSDK
+        SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/third_party/MAVSDK"
         CMAKE_ARGS
             -DCMAKE_INSTALL_PREFIX=${CMAKE_CURRENT_BINARY_DIR}/build_mavsdk/install
-            -DENABLE_MAVLINK_PASSTHROUGH=1
-            GIT_REPOSITORY https://github.com/mavlink/MAVSDK.git
-            GIT_TAG 730a0ef51d248a4c3f39fdb011ed07d28a7c37aa
+           -DBUILD_TESTS=OFF
+        GIT_REPOSITORY https://github.com/mavlink/MAVSDK.git
+        GIT_TAG 5a6d5e3a04c526db49500902190b29b46a786d85
         BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/build_mavsdk
         INSTALL_COMMAND ${CMAKE_COMMAND} --build . --target install
         )

--- a/cmake/dependencies.cmake
+++ b/cmake/dependencies.cmake
@@ -11,7 +11,7 @@ else()
         SOURCE_DIR "${CMAKE_CURRENT_BINARY_DIR}/third_party/MAVSDK"
         CMAKE_ARGS
             -DCMAKE_INSTALL_PREFIX=${CMAKE_CURRENT_BINARY_DIR}/build_mavsdk/install
-           -DBUILD_TESTS=OFF
+            -DBUILD_TESTS=OFF
         GIT_REPOSITORY https://github.com/mavlink/MAVSDK.git
         GIT_TAG 5a6d5e3a04c526db49500902190b29b46a786d85
         BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/build_mavsdk
@@ -31,12 +31,15 @@ endif()
 ExternalProject_Add(third_party_yaml
     SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/third_party/yaml-cpp
     CMAKE_ARGS
-        -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}
-       -DYAML_CPP_BUILD_TESTS=OFF
+        -DCMAKE_INSTALL_PREFIX=${CMAKE_CURRENT_BINARY_DIR}/build_yaml/install
+        -DYAML_CPP_BUILD_TESTS=OFF
     GIT_REPOSITORY https://github.com/jbeder/yaml-cpp.git
     GIT_TAG yaml-cpp-0.6.3
     BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/build_yaml
-    INSTALL_COMMAND ""
+    INSTALL_COMMAND ${CMAKE_COMMAND} --build . --target install
     )
-include_directories(${CMAKE_CURRENT_BINARY_DIR}/third_party/yaml-cpp/include)
+include_directories(${CMAKE_CURRENT_BINARY_DIR}/build_yaml/install/include)
+# Whichever path works :)
+link_directories(${CMAKE_CURRENT_BINARY_DIR}/build_yaml/install/lib)
+link_directories(${CMAKE_CURRENT_BINARY_DIR}/build_yaml/install/lib64)
 list(APPEND dependencies third_party_yaml)

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -37,9 +37,9 @@ int main(int argc, char** argv)
     // For Serial : serial:///path/to/serial/dev[:baudrate]
     connection_result = mavsdk.add_any_connection(connection_url);
 
-    if (connection_result != ConnectionResult::SUCCESS) {
+    if (connection_result != ConnectionResult::Success) {
         std::cout << error_console_text
-                  << "Connection failed: " << connection_result_str(connection_result)
+                  << "Connection failed: " << connection_result
                   << normal_console_text << std::endl;
         return 1;
     }
@@ -48,20 +48,20 @@ int main(int argc, char** argv)
     auto prom = std::make_shared<std::promise<void>>();
     std::future<void> fut = prom->get_future();
 
-    System& system = mavsdk.system();
-    system.register_component_discovered_callback([prom, &system](ComponentType component_type) {
-        std::cout << normal_console_text << "Discovered a component with type "
-                  << unsigned(component_type) << std::endl;
+    mavsdk.subscribe_on_new_system([prom, &mavsdk]() {
+        std::cout << normal_console_text << "Discovered a system\n";
         try {
-            prom->set_value();
             // We only need to receive this once, now we can unsubscribe again.
-            system.register_component_discovered_callback(nullptr);
+            mavsdk.subscribe_on_new_system(nullptr);
+            prom->set_value();
         } catch (const std::future_error& e) {
             // Ignore if prom is set multiple times which can happen even though we're
             // unsubscribing straightaway if the callback triggers two times in quick
             // succession.
         }
     });
+
+    auto system = mavsdk.systems()[0];
 
     if (fut.wait_for(std::chrono::seconds(2)) == std::future_status::timeout) {
         std::cout << error_console_text << "No MAVLink component found" << normal_console_text

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -1,4 +1,6 @@
 
+#include <mavsdk/mavsdk.h>
+
 #include <algorithm>
 #include <chrono>
 #include <cstdint>
@@ -7,8 +9,6 @@
 #include <iostream>
 #include <thread>
 #include <vector>
-
-#include <mavsdk/mavsdk.h>
 
 #include "tests/base.h"
 
@@ -38,8 +38,7 @@ int main(int argc, char** argv)
     connection_result = mavsdk.add_any_connection(connection_url);
 
     if (connection_result != ConnectionResult::Success) {
-        std::cout << error_console_text
-                  << "Connection failed: " << connection_result
+        std::cout << error_console_text << "Connection failed: " << connection_result
                   << normal_console_text << std::endl;
         return 1;
     }

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -75,7 +75,7 @@ int main(int argc, char** argv)
     std::vector<bool> fails(tests_node.size(), false);
 
     // Run the test(s)
-    tests::Context context{.system = system};
+    tests::Context context{system};
     unsigned i = 0;
     for (auto test_node : tests_node) {
         std::string test_name = test_node["name"].as<std::string>();

--- a/src/tests/base.h
+++ b/src/tests/base.h
@@ -2,7 +2,6 @@
 
 #include <mavsdk/mavsdk.h>
 #include <yaml-cpp/yaml.h>
-#include "streamable.h"
 
 #include <exception>
 #include <iostream>
@@ -11,6 +10,8 @@
 #include <string>
 #include <type_traits>
 #include <utility>
+
+#include "streamable.h"
 
 template <typename T>
 inline void unused(T t)

--- a/src/tests/base.h
+++ b/src/tests/base.h
@@ -25,7 +25,7 @@ namespace tests
  * Data that is supplied to each test
  */
 struct Context {
-    mavsdk::System& system;
+    std::shared_ptr<mavsdk::System> system;
     // TODO: logging, ...
 };
 

--- a/src/tests/camera_settings.cpp
+++ b/src/tests/camera_settings.cpp
@@ -1,5 +1,6 @@
 
 #include "camera_settings.h"
+
 #include <algorithm>
 #include <chrono>
 #include <future>
@@ -36,9 +37,10 @@ mavsdk::Camera::SettingOptions CameraSettings::getPossibleSettingOptions()
     std::cout << "Checking if " << _config.param_name << " is a possible setting." << std::endl;
     auto setting_options = _camera.possible_setting_options();
 
-    const auto found =
-        std::find_if(std::begin(setting_options), std::end(setting_options),
-                    [this](const Camera::SettingOptions& setting_options) { return (setting_options.setting_id == _config.param_name); });
+    const auto found = std::find_if(std::begin(setting_options), std::end(setting_options),
+                                    [this](const Camera::SettingOptions& setting_options) {
+                                        return (setting_options.setting_id == _config.param_name);
+                                    });
 
     ASSERT_TRUE(found != setting_options.end());
 

--- a/src/tests/camera_settings.cpp
+++ b/src/tests/camera_settings.cpp
@@ -19,56 +19,56 @@ CameraSettings::CameraSettings(const Context& context) : TestBase(context), _cam
 void CameraSettings::run()
 {
     selectCamera();
-    getPossibleSettingOptions();
-    getAndSetPossibleOptions();
+    const auto setting_options = getPossibleSettingOptions();
+    getAndSetPossibleOptions(setting_options);
 }
 
 void CameraSettings::selectCamera()
 {
-    _camera.select_camera(_config.camera_id);
+    // FIXME: looks like this interface is gone.
+    //_camera.select_camera(_config.camera_id);
     // FIXME: we need to let the camera plugin initialize after this.
     std::this_thread::sleep_for(std::chrono::seconds(3));
 }
 
-void CameraSettings::getPossibleSettingOptions()
+mavsdk::Camera::SettingOptions CameraSettings::getPossibleSettingOptions()
 {
     std::cout << "Checking if " << _config.param_name << " is a possible setting." << std::endl;
-    std::vector<std::string> settings;
-    _camera.get_possible_setting_options(settings);
+    auto setting_options = _camera.possible_setting_options();
 
-    const bool found_setting =
-        std::any_of(std::begin(settings), std::end(settings),
-                    [this](const std::string& setting) { return (setting == _config.param_name); });
+    const auto found =
+        std::find_if(std::begin(setting_options), std::end(setting_options),
+                    [this](const Camera::SettingOptions& setting_options) { return (setting_options.setting_id == _config.param_name); });
 
-    ASSERT_TRUE(found_setting);
+    ASSERT_TRUE(found != setting_options.end());
+
+    return *found;
 }
 
-void CameraSettings::getAndSetPossibleOptions()
+void CameraSettings::getAndSetPossibleOptions(const mavsdk::Camera::SettingOptions& setting_options)
 {
-    Camera::Option initial_option = getOption(_config.param_name);
+    mavsdk::Camera::Setting setting;
+    setting.setting_id = _config.param_name;
 
-    bool is_range = _camera.is_setting_range(_config.param_name);
-    ASSERT_EQ(_config.param_is_range, is_range);
+    const auto initial_settings = _camera.get_setting(setting);
 
-    if (is_range) {
-        setSomeRangeOptions();
+    ASSERT_EQ(initial_settings.first, mavsdk::Camera::Result::Success);
+
+    ASSERT_EQ(_config.param_is_range, initial_settings.second.is_range);
+
+    if (initial_settings.second.is_range) {
+        setSomeRangeOptions(setting_options);
     } else {
-        setAllEnumOptions();
+        setAllEnumOptions(setting_options);
     }
 
     std::cout << "Resetting setting to initial option." << std::endl;
-    setOption(_config.param_name, initial_option);
+    setOption(_config.param_name, initial_settings.second.option);
 }
 
-void CameraSettings::setAllEnumOptions()
+void CameraSettings::setAllEnumOptions(const mavsdk::Camera::SettingOptions& setting_options)
 {
-    Camera::Option initial_option = getOption(_config.param_name);
-
-    std::cout << "Getting possible settings for " << _config.param_name << std::endl;
-    std::vector<Camera::Option> options;
-    _camera.get_possible_options(_config.param_name, options);
-
-    for (const auto& option : options) {
+    for (const auto& option : setting_options.options) {
         std::cout << "- " << option.option_description << " (" << option.option_id << ")"
                   << std::endl;
         setOption(_config.param_name, option);
@@ -77,18 +77,14 @@ void CameraSettings::setAllEnumOptions()
     }
 }
 
-void CameraSettings::setSomeRangeOptions()
+void CameraSettings::setSomeRangeOptions(const mavsdk::Camera::SettingOptions& setting_options)
 {
-    std::cout << "Getting range properties for " << _config.param_name << std::endl;
-    std::vector<Camera::Option> options;
-    _camera.get_possible_options(_config.param_name, options);
-
-    ASSERT_TRUE(options.size() >= 2);
-    ASSERT_TRUE(options.size() <= 3);
+    ASSERT_TRUE(setting_options.options.size() >= 2);
+    ASSERT_TRUE(setting_options.options.size() <= 3);
 
     // FIXME: dealing with this interface is not great and needs improvement.
 
-    const auto minimum = std::atoi(options[0].option_id.c_str());
+    const auto minimum = std::atoi(setting_options.options[0].option_id.c_str());
     {
         Camera::Option new_option;
         new_option.option_id = std::to_string(minimum);
@@ -97,7 +93,7 @@ void CameraSettings::setSomeRangeOptions()
         EXPECT_EQ(set_option, new_option);
     }
 
-    const auto maximum = std::atoi(options[1].option_id.c_str());
+    const auto maximum = std::atoi(setting_options.options[1].option_id.c_str());
     {
         Camera::Option new_option;
         new_option.option_id = std::to_string(maximum);
@@ -108,7 +104,7 @@ void CameraSettings::setSomeRangeOptions()
 
     EXPECT_TRUE(maximum > minimum);
 
-    if (options.size() == 2) {
+    if (setting_options.options.size() == 2) {
         // We are only given min and max, no interval so any increment should work.
         const auto center = (maximum - minimum) / 2 + minimum;
         Camera::Option new_option;
@@ -118,7 +114,7 @@ void CameraSettings::setSomeRangeOptions()
         EXPECT_EQ(set_option.option_id, std::to_string(center));
 
     } else {
-        const auto interval = std::atoi(options[2].option_id.c_str());
+        const auto interval = std::atoi(setting_options.options[2].option_id.c_str());
 
         EXPECT_EQ((maximum - minimum) % interval, 0);
 
@@ -135,23 +131,22 @@ void CameraSettings::setSomeRangeOptions()
 Camera::Option CameraSettings::getOption(const std::string& setting_id)
 {
     std::cout << "Getting option for " << setting_id << std::endl;
-    Camera::Option option;
-    ASSERT_EQ(_camera.get_option(setting_id, option), Camera::Result::SUCCESS);
-    return option;
+
+    Camera::Setting setting;
+    setting.setting_id = setting_id;
+    const auto result = _camera.get_setting(setting);
+    ASSERT_EQ(result.first, Camera::Result::Success);
+    return result.second.option;
 }
 
 void CameraSettings::setOption(const std::string& setting_id, const Camera::Option& option)
 {
     std::cout << "Setting " << setting_id << " to " << option.option_id << std::endl;
-    std::promise<Camera::Result> prom;
-    std::future<Camera::Result> fut = prom.get_future();
-    _camera.set_option_async([&prom](Camera::Result set_result) { prom.set_value(set_result); },
-                             setting_id, option);
-    auto fut_ret = fut.wait_for(std::chrono::seconds(2));
-    EXPECT_EQ(fut_ret, std::future_status::ready);
-    if (fut_ret == std::future_status::ready) {
-        EXPECT_EQ(fut.get(), Camera::Result::SUCCESS);
-    }
+    Camera::Setting setting;
+    setting.setting_id = setting_id;
+    setting.option = option;
+    auto result = _camera.set_setting(setting);
+    EXPECT_EQ(result, Camera::Result::Success);
 }
 
 }  // namespace tests

--- a/src/tests/camera_settings.h
+++ b/src/tests/camera_settings.h
@@ -1,9 +1,9 @@
 #pragma once
 
-#include "base.h"
-
 #include <mavsdk/mavsdk.h>
 #include <mavsdk/plugins/camera/camera.h>
+
+#include "base.h"
 
 namespace tests
 {

--- a/src/tests/camera_settings.h
+++ b/src/tests/camera_settings.h
@@ -37,10 +37,10 @@ protected:
 
 private:
     void selectCamera();
-    void getPossibleSettingOptions();
-    void getAndSetPossibleOptions();
-    void setAllEnumOptions();
-    void setSomeRangeOptions();
+    mavsdk::Camera::SettingOptions getPossibleSettingOptions();
+    void getAndSetPossibleOptions(const mavsdk::Camera::SettingOptions& setting_options);
+    void setAllEnumOptions(const mavsdk::Camera::SettingOptions& setting_options);
+    void setSomeRangeOptions(const mavsdk::Camera::SettingOptions& setting_options);
 
     mavsdk::Camera::Option getOption(const std::string& setting_id);
     void setOption(const std::string& setting_id, const mavsdk::Camera::Option& option);

--- a/src/tests/mission.cpp
+++ b/src/tests/mission.cpp
@@ -1,5 +1,6 @@
 
 #include "mission.h"
+
 #include <future>
 #include <iostream>
 #include <memory>
@@ -107,8 +108,7 @@ void Mission::compareMissions(const mavsdk::Mission::MissionPlan& plan_a,
         return;
     }
 
-    for (std::size_t i = 0; i < plan_a.mission_items.size();
-         ++i) {
+    for (std::size_t i = 0; i < plan_a.mission_items.size(); ++i) {
         EXPECT_EQ(plan_a.mission_items[i], plan_b.mission_items[i]);
     }
 }

--- a/src/tests/mission.cpp
+++ b/src/tests/mission.cpp
@@ -20,12 +20,13 @@ Mission::Mission(const Context& context)
 {
 }
 
-shared_ptr<mavsdk::MissionItem> Mission::makeMissionItem(double latitude_deg, double longitude_deg,
-                                                         float relative_altitude_m)
+mavsdk::Mission::MissionItem Mission::makeMissionItem(double latitude_deg, double longitude_deg,
+                                                      float relative_altitude_m)
 {
-    shared_ptr<mavsdk::MissionItem> new_item = make_shared<mavsdk::MissionItem>();
-    new_item->set_position(latitude_deg, longitude_deg);
-    new_item->set_relative_altitude(relative_altitude_m);
+    mavsdk::Mission::MissionItem new_item{};
+    new_item.latitude_deg = latitude_deg;
+    new_item.longitude_deg = longitude_deg;
+    new_item.relative_altitude_m = relative_altitude_m;
     return new_item;
 }
 
@@ -41,88 +42,74 @@ void Mission::run()
 
 void Mission::uploadDownloadCompare()
 {
-    std::vector<std::shared_ptr<mavsdk::MissionItem>> items = assembleMissionItems();
+    const auto plan = assembleMissionPlan();
 
-    uploadMission(items);
-    std::vector<std::shared_ptr<mavsdk::MissionItem>> downloaded_items = downloadMission();
+    uploadMission(plan);
+    const auto downloaded_plan = downloadMission();
 
-    compareMissions(items, downloaded_items);
+    compareMissions(plan, downloaded_plan);
 }
 
 void Mission::eraseMission()
 {
-    // TODO: presumably the Dronecode SDK should expose a function to do `MISSION_CLEAR_ALL`
-    //       instead of uploading an empty mission list.
-
-    // This doesn't seem to work for now.
-    // std::vector<std::shared_ptr<mavsdk::MissionItem>> no_items{};
-    // uploadMission(no_items);
+    cout << "Clearing mission..." << endl;
+    auto result = _mission.clear_mission();
+    ASSERT_EQ(result, mavsdk::Mission::Result::Success);
 }
 
-std::vector<std::shared_ptr<mavsdk::MissionItem>> Mission::assembleMissionItems()
+mavsdk::Mission::MissionPlan Mission::assembleMissionPlan()
 {
     cout << "Number of waypoints: " << _config.num_waypoints << endl;
 
-    std::vector<std::shared_ptr<mavsdk::MissionItem>> items{};
+    mavsdk::Mission::MissionPlan plan;
 
     for (int i = 0; i < _config.num_waypoints; ++i) {
         float altitude = 10.F + (float)i;
         double latitude = 47.398170327054473 + (double)i * 1e-5;
-        items.push_back(makeMissionItem(latitude, 8.5456490218639658, altitude));
+        plan.mission_items.push_back(makeMissionItem(latitude, 8.5456490218639658, altitude));
     }
 
-    return items;
+    return plan;
 }
 
-void Mission::uploadMission(const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items)
+void Mission::uploadMission(const mavsdk::Mission::MissionPlan& plan)
 {
     cout << "Uploading mission..." << endl;
     promise<mavsdk::Mission::Result> prom{};
     auto fut = prom.get_future();
 
     _mission.upload_mission_async(
-        items, [&prom](mavsdk::Mission::Result result) { prom.set_value(result); });
+        plan, [&prom](mavsdk::Mission::Result result) { prom.set_value(result); });
 
     // wait until uploaded
     const mavsdk::Mission::Result result = fut.get();
-    ASSERT_EQ(result, mavsdk::Mission::Result::SUCCESS);
+    ASSERT_EQ(result, mavsdk::Mission::Result::Success);
 }
 
-std::vector<std::shared_ptr<mavsdk::MissionItem>> Mission::downloadMission()
+const mavsdk::Mission::MissionPlan Mission::downloadMission()
 {
     cout << "Downloading mission..." << endl;
-    promise<pair<mavsdk::Mission::Result, std::vector<std::shared_ptr<mavsdk::MissionItem>>>>
-        prom{};
-    auto fut = prom.get_future();
 
-    _mission.download_mission_async(
-        [&prom](mavsdk::Mission::Result result,
-                const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items) {
-            prom.set_value(make_pair(result, items));
-        });
-
-    auto value = fut.get();
-    const mavsdk::Mission::Result& result = value.first;
-    const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items = value.second;
+    auto result = _mission.download_mission();
 
     // wait until uploaded
-    ASSERT_EQ(result, mavsdk::Mission::Result::SUCCESS);
+    ASSERT_EQ(result.first, mavsdk::Mission::Result::Success);
 
-    return items;
+    return result.second;
 }
 
-void Mission::compareMissions(const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items_a,
-                              const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items_b)
+void Mission::compareMissions(const mavsdk::Mission::MissionPlan& plan_a,
+                              const mavsdk::Mission::MissionPlan& plan_b)
 {
-    EXPECT_EQ(items_a.size(), items_b.size());
+    EXPECT_EQ(plan_a.mission_items.size(), plan_b.mission_items.size());
 
-    if (items_a.size() != items_b.size()) {
+    if (plan_a.mission_items.size() != plan_b.mission_items.size()) {
         return;
     }
 
-    for (std::vector<std::shared_ptr<mavsdk::MissionItem>>::size_type i = 0; i < items_a.size();
+    for (std::size_t i = 0; i < plan_a.mission_items.size();
          ++i) {
-        EXPECT_EQ(*(items_a[i]), *(items_b[i]));
+        EXPECT_EQ(plan_a.mission_items[i], plan_b.mission_items[i]);
     }
 }
 

--- a/src/tests/mission.h
+++ b/src/tests/mission.h
@@ -38,14 +38,14 @@ protected:
 private:
     void uploadDownloadCompare();
     void eraseMission();
-    std::shared_ptr<mavsdk::MissionItem> makeMissionItem(double latitude_deg, double longitude_deg,
-                                                         float relative_altitude_m);
+    mavsdk::Mission::MissionItem makeMissionItem(double latitude_deg, double longitude_deg,
+                                                 float relative_altitude_m);
 
-    std::vector<std::shared_ptr<mavsdk::MissionItem>> assembleMissionItems();
-    void uploadMission(const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items);
-    std::vector<std::shared_ptr<mavsdk::MissionItem>> downloadMission();
-    void compareMissions(const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items_a,
-                         const std::vector<std::shared_ptr<mavsdk::MissionItem>>& items_b);
+    mavsdk::Mission::MissionPlan assembleMissionPlan();
+    void uploadMission(const mavsdk::Mission::MissionPlan& plan);
+    const mavsdk::Mission::MissionPlan downloadMission();
+    void compareMissions(const mavsdk::Mission::MissionPlan& plan_a,
+                         const mavsdk::Mission::MissionPlan& plan_b);
 
     void dropMessages(float ratio);
     bool shouldDropMissionMessage(const mavlink_message_t& message, float ratio);

--- a/src/tests/mission.h
+++ b/src/tests/mission.h
@@ -1,11 +1,11 @@
 #pragma once
 
-#include "base.h"
-#include "lossy_link.h"
-
 #include <mavsdk/mavsdk.h>
 #include <mavsdk/plugins/mavlink_passthrough/mavlink_passthrough.h>
 #include <mavsdk/plugins/mission/mission.h>
+
+#include "base.h"
+#include "lossy_link.h"
 
 namespace tests
 {

--- a/src/tests/param_change.cpp
+++ b/src/tests/param_change.cpp
@@ -23,12 +23,12 @@ void ParamChange::setParam()
         int32_t set_value = long(std::round(_config.set_value));
         cout << "Resetting int32 param " << _config.name << " to " << set_value << endl;
         Param::Result set_result = _param.set_param_int(_config.name, set_value);
-        EXPECT_EQ(set_result, Param::Result::SUCCESS);
+        EXPECT_EQ(set_result, Param::Result::Success);
 
     } else if (isFloat()) {
         cout << "Resetting float param " << _config.name << " to " << _config.set_value << endl;
         Param::Result set_result = _param.set_param_float(_config.name, _config.set_value);
-        EXPECT_EQ(set_result, Param::Result::SUCCESS);
+        EXPECT_EQ(set_result, Param::Result::Success);
 
     } else {
         throw std::invalid_argument("Unknown param type");
@@ -41,12 +41,12 @@ void ParamChange::resetParam()
         int32_t reset_value = long(std::round(_config.reset_value));
         cout << "Setting int32 param " << _config.name << " to " << reset_value << endl;
         Param::Result reset_result = _param.set_param_int(_config.name, reset_value);
-        EXPECT_EQ(reset_result, Param::Result::SUCCESS);
+        EXPECT_EQ(reset_result, Param::Result::Success);
 
     } else if (isFloat()) {
         cout << "Setting float param " << _config.name << " to " << _config.reset_value << endl;
         Param::Result set_result = _param.set_param_float(_config.name, _config.reset_value);
-        EXPECT_EQ(set_result, Param::Result::SUCCESS);
+        EXPECT_EQ(set_result, Param::Result::Success);
 
     } else {
         throw std::invalid_argument("Unknown param type");
@@ -60,14 +60,14 @@ void ParamChange::verifySetParam()
         cout << "Checking set param " << _config.name << " is " << set_value << endl;
         const std::pair<Param::Result, int> get_result = _param.get_param_int(_config.name);
 
-        EXPECT_EQ(get_result.first, Param::Result::SUCCESS);
+        EXPECT_EQ(get_result.first, Param::Result::Success);
         EXPECT_EQ(get_result.second, set_value);
 
     } else if (isFloat()) {
         cout << "Checking set param " << _config.name << " is " << _config.set_value << endl;
         const std::pair<Param::Result, float> get_result = _param.get_param_float(_config.name);
 
-        EXPECT_EQ(get_result.first, Param::Result::SUCCESS);
+        EXPECT_EQ(get_result.first, Param::Result::Success);
         EXPECT_EQ(get_result.second, _config.set_value);
 
     } else {
@@ -82,14 +82,14 @@ void ParamChange::verifyResetParam()
         cout << "Checking reset param " << _config.name << " is " << reset_value << endl;
         const std::pair<Param::Result, int> get_result = _param.get_param_int(_config.name);
 
-        EXPECT_EQ(get_result.first, Param::Result::SUCCESS);
+        EXPECT_EQ(get_result.first, Param::Result::Success);
         EXPECT_EQ(get_result.second, reset_value);
 
     } else if (isFloat()) {
         cout << "Checking reset param " << _config.name << " is " << _config.reset_value << endl;
         const std::pair<Param::Result, float> get_result = _param.get_param_float(_config.name);
 
-        EXPECT_EQ(get_result.first, Param::Result::SUCCESS);
+        EXPECT_EQ(get_result.first, Param::Result::Success);
         EXPECT_EQ(get_result.second, _config.reset_value);
 
     } else {

--- a/src/tests/param_change.cpp
+++ b/src/tests/param_change.cpp
@@ -1,5 +1,6 @@
 
 #include "param_change.h"
+
 #include <cmath>
 #include <cstdint>
 #include <future>

--- a/src/tests/param_change.h
+++ b/src/tests/param_change.h
@@ -1,9 +1,9 @@
 #pragma once
 
-#include "base.h"
-
 #include <mavsdk/mavsdk.h>
 #include <mavsdk/plugins/param/param.h>
+
+#include "base.h"
 
 namespace tests
 {

